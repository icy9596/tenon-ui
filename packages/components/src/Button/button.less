.btn {
  text-align: center;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 4px 10px;
  background-color: transparent;
  cursor: pointer;
  user-select: none;
  transition: border-color 0.5s, background-color 0.5s;

  &-default {
    border-color: #ccc;

    &:hover {
      border-color: green;
    }
  }

  &-primary {
    background-color: green;
    color: white;

    &:hover {
      background-color: #52c41a;
    }
  }
}
