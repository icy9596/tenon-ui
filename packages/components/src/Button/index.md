---
title: Button
demo:
  cols: 2
nav:
  title: 组件
group:
  title: 通用
---

# Button 按钮

按钮用于开始一个即时操作。

## 代码演示

<code src="./demo/type.tsx">按钮类型</code>
<code src="./demo/size.tsx">按钮尺寸</code>
