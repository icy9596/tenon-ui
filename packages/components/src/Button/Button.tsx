import classnames from 'classnames';

import './button.less';

export type ButtonType = 'primary' | 'default';

export interface ButtonProps {
  type?: ButtonType;
  onClick?: () => void;
  children: string;
}

const Button = ({
  type = 'default',
  onClick,
  children,
}: ButtonProps): JSX.Element => {
  const className = classnames(
    'btn',
    type === 'default' ? 'btn-default' : 'btn-primary',
  );

  const handleClick = () => {
    onClick?.();
  };

  return (
    <button className={className} type="button" onClick={handleClick}>
      {children}
    </button>
  );
};

export default Button;
