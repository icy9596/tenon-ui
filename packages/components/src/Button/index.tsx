import Button from './Button';
export type { ButtonProps, ButtonType } from './Button';

export default Button;
