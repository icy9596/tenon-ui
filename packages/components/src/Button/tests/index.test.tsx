import { fireEvent, render, screen } from '@testing-library/react';
import Button from '../Button';

const btnText = 'test_btn';

describe('Button', () => {
  it('render', () => {
    const { container } = render(<Button>{btnText}</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('mount', () => {
    render(<Button>{btnText}</Button>);
    expect(screen.getByText(btnText)).toBeInTheDocument();
  });

  it('type', () => {
    // defualt
    const { rerender } = render(<Button>{btnText}</Button>);
    expect(screen.getByText(btnText)).toHaveClass('btn-default');

    // primary
    rerender(<Button type="primary">{btnText}</Button>);
    expect(screen.getByText(btnText)).toHaveClass('btn-primary');
  });

  it('onClick', () => {
    const handleClick = jest.fn();
    const { rerender } = render(
      <Button onClick={handleClick}>{btnText}</Button>,
    );
    fireEvent.click(screen.getByText(btnText));
    expect(handleClick).toBeCalled();
    expect(handleClick).toBeCalledTimes(1);

    // no call
    rerender(<Button>{btnText}</Button>);
    fireEvent.click(screen.getByText(btnText));
  });
});
