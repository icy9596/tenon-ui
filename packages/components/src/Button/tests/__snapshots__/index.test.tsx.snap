// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render 1`] = `
<button
  class="btn btn-default"
  type="button"
>
  test_btn
</button>
`;
